File: csharp/TransactionAnalyzer.Lib/FileTransactionLoader.cs
using System;
using System.Globalization;
using System.IO;
using System.Linq;

namespace TransactionAnalyzer.Lib
{
    public class FileTransactionLoader : ITransactionLoader
    {
        private readonly string _csvFilePath;

        public FileTransactionLoader(string csvFilePath)
        {
            _csvFilePath = csvFilePath;
        }

        public Transaction[] Load()
        {
            var lines = File.ReadAllLines(_csvFilePath);
            return lines.Skip(1).Select(BuildTransaction).ToArray();
        }

        private Transaction BuildTransaction(string line)
        {
            Console.WriteLine("reading line:" + line);
            // Split into array
            var fields = line.Split(",");
            Console.WriteLine("Split into fields:" + fields.Length);

            //try to parse the fields
            Enum.TryParse(fields[4].Trim(), out TransactionType parsedType);
            DateTime.TryParseExact(fields[1].Trim(), "dd/MM/yyyy HH:mm:ss", CultureInfo.InvariantCulture,
                DateTimeStyles.None, out var parsedTransactedAt);
            return new Transaction{
                Id = fields[0].Trim(),
                TransactedAt = parsedTransactedAt,
                Amount = Convert.ToDecimal(fields[2].Trim()),
                MerchantName = fields[3].Trim(),
                Type = parsedType,
                RelatedTransactionId = fields[5]?.Trim()
            };
        }
    }
}

File: transaction-analyzer/node-typescript/src/file-transaction-loader.ts
import Big from 'big.js';
import { DateTimeFormatter, LocalDateTime } from '@js-joda/core';
import { TransactionType } from './transaction-type.enum';
import { Transaction } from './transaction.interface';
import { TransactionLoader } from './transaction-loader.interface';
import { readFileSync } from 'node:fs';
export class DefaultTransactionLoader implements TransactionLoader {
  constructor(private file: string) {}

  load(): Transaction[] {
    const lines = readFileSync(this.file, 'utf-8').split('\n').filter(Boolean);
    return lines.slice(1).map((line: string) => this.buildTransaction(line));
  }

  buildTransaction(line: string): Transaction {
    console.log('reading line:' + line);
    const fields = line.split(',');
    console.log('fields: ' + JSON.stringify(fields));
    return {
      id: fields[0].trim(),
      transactedAt: LocalDateTime.parse(
        fields[1].trim(),
        DateTimeFormatter.ofPattern('dd/MM/yyyy HH:mm:ss')
      ),
      amount: Big(fields[2].trim()),
      merchantName: fields[3].trim(),
      type: TransactionType[fields[4].trim()],
      relatedTransactionId: fields[5].trim()
    };
  }
}
